src/details/ArborX_DetailsTreeTraversal.hpp

```diff
@@ -450,6 +450,146 @@ struct TreeTraversal<BVH, Predicates, Callback, NearestPredicateTag>
   }
 };
 
+template <class BVH, class Predicates, class Callback>
+struct TreeTraversal<BVH, Predicates, Callback,
+                     Experimental::OrderedNearestPredicateTag>
+{
+  BVH _bvh;
+  Predicates _predicates;
+  Callback _callback;
+
+  using Access = AccessTraits<Predicates, PredicatesTag>;
+
+  template <class ExecutionSpace>
+  TreeTraversal(ExecutionSpace const &space, BVH const &bvh,
+                Predicates const &predicates, Callback const &callback)
+      : _bvh{bvh}
+      , _predicates{predicates}
+      , _callback{callback}
+  {
+    if (_bvh.empty())
+    {
+      // do nothing
+    }
+    else if (_bvh.size() == 1)
+    {
+      Kokkos::parallel_for(
+          "ArborX::Experimental::TreeTraversal::OrderedNearestPredicate"
+          "degenerated_one_leaf_tree",
+          Kokkos::RangePolicy<ExecutionSpace, OneLeafTree>(
+              space, 0, Access::size(predicates)),
+          *this);
+    }
+    else
+    {
+      Kokkos::parallel_for(
+          "ArborX::Experimental::TreeTraversal::OrderedNearestPredicate",
+          Kokkos::RangePolicy<ExecutionSpace>(space, 0,
+                                              Access::size(predicates)),
+          *this);
+    }
+  }
+
+  struct OneLeafTree
+  {};
+
+  KOKKOS_FUNCTION void operator()(OneLeafTree, int queryIndex) const
+  {
+    auto const &predicate = Access::get(_predicates, queryIndex);
+    auto const root = HappyTreeFriends::getRoot(_bvh);
+    auto const &root_bounding_volume =
+        HappyTreeFriends::getBoundingVolume(_bvh, root);
+    constexpr auto inf = KokkosExt::ArithmeticTraits::infinity<float>::value;
+    if (distance(getGeometry(predicate), root_bounding_volume) != inf)
+    {
+      _callback(predicate, 0);
+    }
+  }
+
+  KOKKOS_FUNCTION void operator()(int queryIndex) const
+  {
+    auto const &predicate = Access::get(_predicates, queryIndex);
+    using ArborX::Details::HappyTreeFriends;
+
+    auto const distance = [geometry = getGeometry(predicate),
+                           bvh = _bvh](int node) {
+      auto const &box = HappyTreeFriends::getBoundingVolume(bvh, node);
+      using Details::distance;
+      return distance(geometry, box);
+    };
+
+    using PairIndexDistance = Kokkos::pair<int, float>;
```

Comment:
If we are trying to generalize this, the second argument should the type returned by the `distance` function. As it can be provided by a user, this will have to be determined and not hardcoded.